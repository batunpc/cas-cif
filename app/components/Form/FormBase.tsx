import Form from "lib/theme/FormWithTheme";
import type { JSONSchema7 } from "json-schema";
import FormComponentProps from "./FormComponentProps";
interface Props extends FormComponentProps {
  schema: JSONSchema7;
  uiSchema: {};
}

const FormBase: React.FC<Props> = ({
  formData,
  onChange,
  onFormErrors,
  schema,
  uiSchema,
}) => {
  return (
    <>
      <Form
        noHtml5Validate
        schema={schema}
        uiSchema={uiSchema}
        formData={formData}
        onChange={(change) => {
          onChange(change.formData);
          onFormErrors(change.errors);
        }}
        omitExtraData
      ></Form>
      <style jsx>{`
        :global(label) {
          font-weight: bold;
        }
      `}</style>
    </>
  );
};

export default FormBase;
