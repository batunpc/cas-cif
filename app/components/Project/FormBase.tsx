import FormBorder from "components/Layout/FormBorder";
import { useEffect, useRef } from "react";
import Form from "lib/theme/service-development-toolkit-form";
import { JSONSchema7 } from "json-schema";
import FormComponentProps from "./FormComponentProps";

interface Props extends FormComponentProps {
  createInitialData: (data: object) => object;
  schema: JSONSchema7;
  uiSchema: object;
}

const FormBase: React.FunctionComponent<Props> = ({
  formData,
  applyChangesFromComponent,
  onFormErrors,
  createInitialData,
  schema,
  uiSchema,
}) => {
  // We restrict the data to only the fields we care about
  // It's important to default to `undefined` so we trigger the errors on rjsf
  const initialData = createInitialData(formData);

  const makeErrorsObject = (errorSchema) => {
    const keys = Object.keys(initialData);
    const returnVal = {};
    keys.forEach((key) => {
      if (errorSchema[key]) {
        returnVal[key] = errorSchema[key];
      } else {
        returnVal[key] = null;
      }
    });
    return returnVal;
  };

  const formRef = useRef();

  useEffect(() => {
    // @ts-ignore
    const { errorSchema } = formRef.current.validate(initialData, schema);
    onFormErrors(makeErrorsObject(errorSchema));
  }, []);

  return (
    <>
      <FormBorder title="Background">
        <Form
          // @ts-ignore
          ref={formRef}
          schema={schema}
          uiSchema={uiSchema}
          formData={initialData}
          onChange={(change) => {
            applyChangesFromComponent(change.formData);
            onFormErrors(makeErrorsObject(change.errorSchema));
          }}
          liveValidate
        ></Form>
      </FormBorder>
    </>
  );
};

export default FormBase;
