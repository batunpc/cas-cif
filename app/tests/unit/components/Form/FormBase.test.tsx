import { render } from "@testing-library/react";
import FormBase from "components/Form/FormBase";
import readOnlyTheme from "lib/theme/ReadOnlyTheme";

const testSchema = {
  type: "object",
  required: ["field"],
  properties: {
    field: {
      type: "string",
    },
  },
};

const testUiSchema = {
  type: {
    "ui:widget": "TextWidget",
  },
};

describe("The FormBase component", () => {
  it("renders a test schema with the default theme", () => {
    const props = {
      schema: testSchema,
      formData: { field: "test" },
      uiSchema: testUiSchema,
    } as any;

    const formUnderTest = render(<FormBase {...props} />);

    expect(formUnderTest.container).toMatchSnapshot();
  });

  it("renders a test schema with the readonly theme", () => {
    const props = {
      schema: testSchema,
      formData: { field: "test" },
      uiSchema: testUiSchema,
      theme: readOnlyTheme,
    } as any;

    const formUnderTest = render(<FormBase {...props} />);

    expect(formUnderTest.container).toMatchSnapshot();
  });

  it("passes the ref through with an HTML form element", () => {
    const props = {
      schema: {},
      uiSchema: {},
    } as any;

    const refUnderTest = jest.fn();

    render(<FormBase {...props} ref={refUnderTest} />);

    expect(refUnderTest).toHaveBeenCalledTimes(1);
    expect(refUnderTest).toHaveBeenCalledWith(
      expect.objectContaining({
        formElement: expect.any(HTMLFormElement),
      })
    );
  });
});
