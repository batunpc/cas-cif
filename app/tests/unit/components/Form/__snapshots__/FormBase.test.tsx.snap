// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FormBase component renders a test schema 1`] = `
<div>
  <form
    class="rjsf"
    novalidate=""
  >
    <div
      class="jsx-2194225068"
    >
      <div
        class="sc-jRQBWg cDwDrH pg-grid-container"
        cols="10"
      >
        <div
          class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
        >
          <div
            class="sc-furwcr sc-pVTFL XpgcI eqAbko pg-grid-col"
            width="100"
          >
            <fieldset
              class="jsx-4178779636"
            >
              <div
                class="jsx-2194225068"
              >
                <div
                  class="sc-crHmcD bKATQK pg-input"
                >
                  <label
                    class="sc-egiyK ewKLoj pg-input-label"
                    for="root_field"
                  >
                    field
                  </label>
                  <input
                    aria-label="field"
                    class="sc-bqiRlB hrAFNs pg-input-input"
                    id="root_field"
                    name="root_field-input"
                    placeholder=""
                    required=""
                    value="test"
                  />
                </div>
                <div
                  class="jsx-2194225068 error-div"
                >
                  <div
                    class="jsx-2194225068"
                  >
                     
                  </div>
                </div>
              </div>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="jsx-2194225068 error-div"
      >
        <div
          class="jsx-2194225068"
        >
           
        </div>
      </div>
    </div>
  </form>
</div>
`;
